internal/timeouts/timeouts.go

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) SUSE Software Solutions Germany GmbH
+// SPDX-License-Idendifier: Apache-2.0
+
+package timeouts
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+const (
+	minimumTimeout = time.Duration(30 * time.Second)
+)
+
+func ContextWithCreateTimeout(ctx context.Context, d *schema.ResourceData) (context.Context, context.CancelFunc) {
+	return contextWithTimeout(ctx, d, schema.TimeoutCreate)
+}
+
+func ContextWithReadTimeout(ctx context.Context, d *schema.ResourceData) (context.Context, context.CancelFunc) {
+	return contextWithTimeout(ctx, d, schema.TimeoutRead)
+}
+
+func ContextWithUpdateTimeout(ctx context.Context, d *schema.ResourceData) (context.Context, context.CancelFunc) {
+	return contextWithTimeout(ctx, d, schema.TimeoutUpdate)
+}
+
+func ContextWithDeleteTimeout(ctx context.Context, d *schema.ResourceData) (context.Context, context.CancelFunc) {
+	return contextWithTimeout(ctx, d, schema.TimeoutDelete)
+}
+
+func ContextWithDefaultTimeout(ctx context.Context, d *schema.ResourceData) (context.Context, context.CancelFunc) {
+	return contextWithTimeout(ctx, d, schema.TimeoutDefault)
+}
+
+func contextWithTimeout(ctx context.Context, d *schema.ResourceData, key string) (context.Context, context.CancelFunc) {
+	var timeout time.Duration
+
+	// If there is already a deadline set, we don't want to set another one. Since context.WithTimeout is essentially
+	// the same as context.WithDeadline(parent, time.Now().Add(timeout)), this would add the new timeout to the existing
+	// one, essentially yielding longer timeouts than anticipated. This happens e.g. when resourceImageRead(..) is
+	// called from resourceImageUpdate(..).
+	_, ok := ctx.Deadline()
+	if ok {
+		return context.WithCancel(ctx)
+	}
+
+	// Enforce a minimum timeout to prevent users from setting a ridiculously low
+	// timeout, which could cause Terraform to become unable to modify resources.
+	if minimumTimeout <= d.Timeout(key) {
+		timeout = d.Timeout(key)
+	} else {
+		timeout = minimumTimeout
+	}
```

Comment:
This seems not work. I have tried set timeout to 1s to create vm image and the execution after 1s not 30s.

Reply:
You're right. This flooring of the timeout never worked and my testing wasn't proper.

I've also realized a few points about the SDK and contexts:

- The context for the `CreateContext`, `ReadContext`, etc. functions of the resources is already populated with the respective timeouts and the default is 20 minutes in the SDK. I've overridden the default to be 2 minutes if no explicit timeout is given. This might be a bit short for some resources, but I think 20 minutes is quite excessive.
This also means quite a bit of code is completely unnecessary as long as the context that is passed into the functions is properly propagated to API calls.

- It's impossible create a derived context with a longer timeout than the parent context, but the context also contains other information, so we can't just instantiate a new context. Therefore it's impossible to modify a given timeout, which is why the minimum timeout never worked.
While this means we can't enforce a minimum timeout, this isn't really going to be a common problem anyways.

Reply:
> While this means we can't enforce a minimum timeout, this isn't really going to be a common problem anyways.

Indeed, now we cant set the default timeout to 2 mins, so this should be fine.